staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+
+	remoteAttsInformer.AddIndexers(map[string]k8scache.IndexFunc{attMACIndexName: attachmentMACAddr})
+
+	remoteAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    ca.onRemoteAttAdded,
+		UpdateFunc: ca.onRemoteAttUpdated,
+		DeleteFunc: ca.onRemoteAttRemoved,
+	})
+
+	remoteAttsInformerStopCh := make(chan struct{})
+	go remoteAttsInformer.Run(aggregateTwoStopChannels(ca.stopCh, remoteAttsInformerStopCh))
+
+	return &vnState{
+		remoteAttsInformer:       remoteAttsInformer,
+		remoteAttsInformerStopCh: remoteAttsInformerStopCh,
+		remoteAttsLister:         remoteAttsLister,
+		namespace:                namespace,
+		localAtts:                make(map[string]struct{}),
+		remoteAtts:               make(map[string]struct{}),
+	}
+}
+
+func (ca *ConnectionAgent) onRemoteAttAdded(obj interface{}) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Remote NetworkAttachments cache for VNI %d: notified of addition of %#+v", att.Spec.VNI, att)
+	attNSN := kosctlrutils.AttNSN(att)
+	ca.addVNI(attNSN, att.Spec.VNI)
+	ca.queue.Add(attNSN)
+}
+
+func (ca *ConnectionAgent) onRemoteAttUpdated(oldObj, newObj interface{}) {
+	oldAtt := oldObj.(*netv1a1.NetworkAttachment)
+	newAtt := newObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Remote NetworkAttachments cache for VNI %d: notified of update from %#+v to %#+v", newAtt.Spec.VNI, oldAtt, newAtt)
+	ca.queue.Add(kosctlrutils.AttNSN(newAtt))
+}
+
+func (ca *ConnectionAgent) onRemoteAttRemoved(obj interface{}) {
+	peeledObj := kosctlrutils.Peel(obj)
+	att := peeledObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Remote NetworkAttachments cache for VNI %d: notified of deletion of %#+v", att.Spec.VNI, att)
+	attNSN := kosctlrutils.AttNSN(att)
+	ca.removeVNI(attNSN, att.Spec.VNI)
+	ca.queue.Add(attNSN)
+}
+
+func (ca *ConnectionAgent) addVNI(nsn k8stypes.NamespacedName, vni uint32) {
+	ca.nsnToVNIsMutex.Lock()
+	defer ca.nsnToVNIsMutex.Unlock()
+	attSeenInVNIs := ca.nsnToVNIs[nsn]
+	if attSeenInVNIs == nil {
+		attSeenInVNIs = uint32set.NewHashUInt32Set()
+		ca.nsnToVNIs[nsn] = attSeenInVNIs
+	}
+	attSeenInVNIs.Add(vni)
+}
+
+func (ca *ConnectionAgent) removeVNI(nsn k8stypes.NamespacedName, vni uint32) {
+	ca.nsnToVNIsMutex.Lock()
+	defer ca.nsnToVNIsMutex.Unlock()
+	attSeenInVNIs := ca.nsnToVNIs[nsn]
+	if attSeenInVNIs == nil {
+		return
+	}
+	attSeenInVNIs.Remove(vni)
+	if attSeenInVNIs.IsEmpty() {
+		delete(ca.nsnToVNIs, nsn)
+	}
+}
+
+func (ca *ConnectionAgent) getAttVNIs(nsn k8stypes.NamespacedName) []uint32 {
+	ca.nsnToVNIsMutex.RLock()
+	defer ca.nsnToVNIsMutex.RUnlock()
+	attSeenInVNIs := ca.nsnToVNIs[nsn]
+	if attSeenInVNIs == nil {
+		return nil
+	}
+	// ? This invocation has a runtime which is O(N), where N = (max_vni - min_vni). We're doing
+	// it while holding a mutex shared with the notif handlers. It could be bad for performance.
+	return attSeenInVNIs.Export()
```

Comment:
Given the usage by the caller, all this function needs to return is a single `uint32`, a boolean indicating whether that int is in the set, and a boolean indicating whether there are additional ints in the set.

Reply:
We might want the whole list for logging purposes. Also, I find it cleaner to return the slice and have the caller infer the information returned by checking the length, rather than returning and using multiple flags.

Reply:
Fixed in 08d7d06. Switched to a function which returns a vni and the number of VNIs found (which carries the same information as the flags you where talking about).